nntrainer/tensor/cl_operations/attention_kernels.cpp

```diff
@@ -0,0 +1,273 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2024 Yash Singh <yash.singh@samsung.com>
+ *
+ * @file	attention_kernels.cpp
+ * @date	28 August 2024
+ * @brief	Common attention OpenCL kernels
+ * @see		https://github.com/nnstreamer/nntrainer
+ * @author	Yash Singh <yash.singh@samsung.com>
+ * @bug		No known bugs except for NYI items
+ *
+ */
+
+#include <attention_kernels.h>
+
+namespace nntrainer {
+std::string rotary_emb_cl_kernel = R"(
+  #pragma OPENCL EXTENSION cl_khr_fp16 : enable
+__kernel void rotary_emb_cl(__global float *input,
+                                      __global float *output,
+                                      __global float *freqs_cos,
+                                      __global float *freqs_sin,
+                                      __global float *cos_,
+                                      __global float *sin_,
+                                      unsigned int batch,
+                                      unsigned int channel,
+                                      unsigned int height,
+                                      unsigned int width,
+                                      unsigned int dim,
+                                      unsigned int half_,
+                                      unsigned int max_timestep,
+                                      unsigned int from) {
+    unsigned int gid = get_global_id(0);
+    unsigned int gws = get_global_size(0);
+
+    __global float *cos_ptr = cos_;
+    __global float *sin_ptr = sin_;
+
+    float value = 0.0f;
+    float transformed_value = 0.0f;
+
+    for (unsigned int b = 0; b < batch; b++) {
+      for (unsigned int c = 0; c < channel; c++) {
+        for (unsigned int h = 0; h < height; h++) {
+          if (from + h < max_timestep) {
+            unsigned idx = (from + h)*dim;
+            for(unsigned int i = idx; i < idx + dim; i++){
+              cos_ptr[i - idx] = freqs_cos[i];
+              sin_ptr[i - idx] = freqs_sin[i];
+            }
+          }
+          for (unsigned int w = 0; w < width; w = w + dim) {
+            for (unsigned int k = 0; k < dim; k++) {
+              unsigned int span = w + k;
+              value = input[b * channel * height * width + c * height * width + h * width + span];
+              if (k < half_) {
+                transformed_value = -1.0f * input[b * channel * height * width + c * height * width + h * width + span + half_];
+              } else {
+                transformed_value = input[b * channel * height * width + c * height * width + h * width + span - half_];
+              }
+              value = value * cos_ptr[k] + transformed_value * sin_ptr[k];
+              // printf("GPU Batch: %u, Height: %u, Channel: %u, Width: %u, K: %u, Span: %u, Value: %f, Transformed Value: %f, cos_ptr[k]: %f, sin_ptr[k]: %f\n",  b, h, c, w, k, span, value, transformed_value, cos_ptr[k], sin_ptr[k]);
```

Comment:
how about removing unnecessary commented print codes

Reply:
Yes, thank you pointing it out. I've removed all the unecessary comments from the code.